lib/jaeger/client/async_reporter.rb

```diff
@@ -34,8 +25,24 @@ def flush
 
       def report(span)
         return if !span.context.sampled? && !span.context.debug?
+
+        init_reporter_thread
```

Comment:
Previously I created the flushing thread in the create method. This allowed easily to test AsyncReporter instance without having to deal with threads. With this solution you'll be creating threads also in the specs and not closing them (until the process itself dies). I'd like to avoid leaking threads even in tests.

What do you think about doing something like this in the `self.create` method:
```
  flusher = PeriodicFlusher.new(interval: interval)
  reporter = new(sender, flusher)
```

and then in `report(span)` method you could do something like: `@flusher.ensure_started`

This allows moving all Thread management related logic to PeriodicFlusher flusher, which allows us easily to test the AsyncReporter class by injecting `instance_spy(PeriodicFlusher)` as flusher. We could also test PeriodicFlusher separately if we want to.

Reply:
I think we shouldn't create the thread in `create` method because if you create the thread in `create` method it will be created if the user has a reference to tracer even if it does not create any span. The thread should be created lazily while adding a span into buffer, by this way we can also recreate the thread in case if it somehow dies.

Why you are not considering `instance_spy(Thread)`? The logic of `PeriodicFlusher` would be quite trivial that can question it's existence. If you think there can be different flushers(Process based, fiber based, thread pool based etc.) then it can make sense otherwise I think it's a bit overkill.

Reply:
I didn't mean creating a thread in `.create`. I meant only creating the `PeriodicFlusher`. Thread will be created when `ensure_started` is called.

I also want to add a flusher that uses span count instead of timer. Also will at some point need to address udp packets that can go over limit which probably needs more advanced flusher.

injecting `instance_spy(Thread)` probably also works, though using more explicit models is imo nicer.

Reply:
> Also will at some point need to address udp packets that can go over limit which probably needs more advanced flusher.

I'm already facing with this problem even for a single span(if I add backtrace of a ruby error as log :()

I'm not completely against of having a separate class for flusher just thinking about how to inject this to `AsyncReporter`. How would you like to set specific flusher implementation while creating the tracer?

Reply:
> I'm already facing with this problem even for a single span(if I add backtrace of a ruby error as log :()

You can use http sender which doesn't have this problem meanwhile if you can.

> How would you like to set specific flusher implementation while creating the tracer?

Imo you can postpone that part. Adding `flusher = ..` to `self.create` is enough. We have a hook and can deal with the actual tracer level injection later (when we add new flushers).

Basically, currently only thing I want, is to avoid leaking threads in the specs. Don't worry about building something bigger.

Reply:
👍 I'll prepare it today or tomorrow

Reply:
Hm. I think lets go with the simplest solution there is atm. You proposed injecting `Thread` so we can do `instance_spy(Thread)` in the tests. I think that's currently the best one. Feels like we're getting a bit into the overengineering field.

Btw, from previous conversation:
> I'm already facing with this problem even for a single span

If even a single span fails, then you might be blocked by some system limits for udp. Jaeger server can handle max packet of ~65000 bytes, but sometimes the system limits are smaller and you have to increase UDP MTU. If the serialized span is over 65K bytes then I don't think we can do anything (for UDP sender). To my knowledge, splitting a serialized span packet is not possible here.

Reply:
> Hm. I think lets go with the simplest solution there is atm. You proposed injecting Thread so we can do instance_spy(Thread) in the tests. I think that's currently the best one. Feels like we're getting a bit into the overengineering field.

It's fine for me.

The packet size was between 60-65k bytes and the error was coming from Ruby's UDP class which indicates I've already reached the maximum size for a UDP message. I've solved it by truncating the backtrace, it's a reasonable solution for now.

Btw I noticed that at some places you are using logger and at rest `warn` which puts to stdout. Do you have a plan to unify this? I'm currently writing a tracer gem on top of this gem therefore I can come up with new ideas in near future.

Reply:
> Do you have a plan to unify this?

Yes. Will unify them under logger at some point.